### PR TITLE
Finetuning + Multi-GPU Support with PyTorch

### PR DESCRIPTION
**Pull request by misha-chertushkin**
## Description
This PR adds finetuning for TimesFM model on PyTorch and also adds the the Multi-GPU training on PyTorch. The framework is designed to be easy to use while providing advanced features for more sophisticated use cases.

## For review
- `notebooks/finetuning_example.py` - client code
- `notebooks/finetuning_example.ipynb` - client code and example how to use finetuning framework
- `timesfm/finetuning_torch.py` - framework code

Framework code contains class `TimesFMFinetuner` which accepts the instance of the model and the `FinetuningConfig`

## Features

- 🚀 Easy-to-use API for finetuning TimesFM models
- 📊 Integration with Weights & Biases for experiment tracking
- 🔧 Configurable training parameters and loss functions
- 📈 Advanced features including:
  - Multi-GPU training support
  - Custom loss functions
  - Data preprocessing utilities
  - Learning rate scheduling
  - Early stopping
  - Model checkpointing
- 📝 Comprehensive logging and metrics tracking

## Testing
Make sure to follow the installation instructions. After you have pyenv env successfully created, just execute `poetry run python notebooks/finetuning_torch.py`. This will:
- Create the TimesFM model on PyTorch
- Load the checkpoint from HuggingFace
- Finetune the model on the yfinance dataset

## Quick Start

```python
import numpy as np
from timesfm import TimesFm, TimesFmHparams
from timesfm_finetuner import FinetuningConfig, TimesFMFinetuner

# Load your time series data
time_series = np.load('my_time_series.npy')

# Initialize TimesFM model
model = TimesFm(
    hparams=TimesFmHparams(
        backend="cuda",
        per_core_batch_size=32,
        horizon_len=128,
        num_layers=50,
    )
)

# Configure finetuning
config = FinetuningConfig(
    batch_size=64,
    num_epochs=50,
    learning_rate=1e-4,
    use_wandb=True  # Enable W&B logging
)

# Create finetuner
finetuner = TimesFMFinetuner(config, model.config)

# Finetune model
results = finetuner.finetune(model, time_series, "my_model")
```

### Multi-GPU Training

```python
  gpu_ids = [0, 1]
  config = FinetuningConfig(
    batch_size=256,
    num_epochs=5,
    learning_rate=3e-5,
    use_wandb=True,
    distributed=True,
    gpu_ids=gpu_ids,
  )
```

## Configuration Options

The `FinetuningConfig` class provides extensive configuration options:

```python
config = FinetuningConfig(
    # Training parameters
    batch_size=32,
    num_epochs=20,
    learning_rate=1e-4,
    weight_decay=0.01,
    
    # Model parameters
    context_length=128,
    horizon_length=32,
    

    # Multi-GPU training
    distributed=True,
    gpu_ids=gpu_ids,
)
```

[WIP]

### CODE REVIEW

File: notebooks/finetuning_example.py
Comment (otakuCandy) [inline notebooks/finetuning_example.py:1]: Hi，nmgvv.
I hope this message finds you well.
Thank you for your amazing work on ‘timesfm’ and for sharing it with the community. I recently came across your pull request and was really excited to explore your code. However, I noticed that some parts of the code have not been uploaded yet, and as a new user, I'm finding it challenging to understand how to utilize the current implementation.
If it's not too much trouble, could you please complete the upload and, if possible, add more comments or examples? I believe it would help not just me but others in the community better understand and benefit from your contributions.
I truly appreciate your effort and understand that preparing and sharing code takes time. Please let me know if there's anything I can do to help with the process.
Thank you again for your hard work, and I look forward to seeing the completed project!

File: notebooks/finetuning_torch.py [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+    Args:
+      batch: List of input tensors.
+
+    Returns:
+      Tuple of (loss, predictions).
+    """
+    x_context, x_padding, freq, x_future = [t.to(self.device, non_blocking=True) for t in batch]
+
+    predictions = self.model(x_context, x_padding.float(), freq)
+    predictions_mean = predictions[..., 0]
```
Comment (rajatsen91) [inline notebooks/finetuning_torch.py]: Can we add support for also tuning the quantiles: see [here](https://github.com/google-research/timesfm/blob/master/src/timesfm/patched_decoder.py#L532) for the loss function in pax

File: notebooks/finetuning_torch.py [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+      Average training loss for the epoch.
+    """
+    self.model.train()
+    total_loss = 0.0
+
+    for batch in train_loader:
+      loss, _ = self._process_batch(batch)
+
+      if self.config.distributed:
+        losses = [torch.zeros_like(loss) for _ in range(dist.get_world_size())]
```
Comment (rajatsen91) [inline notebooks/finetuning_torch.py]: I am not sure I understand why we are doing `torch.zeros_like(loss)` here? Should we not get a set of actual losses from each distributed backprop ?

File: notebooks/finetuning_torch.py [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+    )
+
+    history = {"train_loss": [], "val_loss": [], "learning_rate": []}
+
+    self.logger.info(f"Starting training for {self.config.num_epochs} epochs...")
+    self.logger.info(f"Training samples: {len(train_dataset)}")
+    self.logger.info(f"Validation samples: {len(val_dataset)}")
+
+    try:
+      for epoch in range(self.config.num_epochs):
```
Comment (rajatsen91) [inline notebooks/finetuning_torch.py]: nit: Is it possible to have train and eval loss logging in between epochs as well if the epoch is very large? This is completely optional.

**Review by otakuCandy — COMMENTED** [review]
When fine-tuning, I noticed that the following messages are repeatedly printed:
TimesFM v1.2.0. See https://github.com/google-research/timesfm/blob/master/README.md for updated APIs.
Loaded Jax TimesFM.
Loaded PyTorch TimesFM.
Could you help me take a look at this issue?

File: notebooks/finetuning_example.py
[the remark this answers is not in the document]
Comment (misha-chertushkin) [reply notebooks/finetuning_example.py:1]: Good catch! Shifted 👍

File: notebooks/finetuning_example.py [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+    return len(self.samples)
+
+  def __getitem__(self, index: int) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+    x_context, x_future = self.samples[index]
+
+    x_context = torch.tensor(x_context, dtype=torch.float32)
+    x_future = torch.tensor(x_future, dtype=torch.float32)
+
+    input_padding = torch.zeros_like(x_context)
+    freq = torch.zeros(1, dtype=torch.long)
```
[the remark this answers is not in the document]
Comment (misha-chertushkin) [reply notebooks/finetuning_example.py:60]: Good suggestion, fixed :)

File: notebooks/finetuning_example.py [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+def get_model(load_weights: bool = False):
+  device = "cuda" if torch.cuda.is_available() else "cpu"
+  repo_id = "google/timesfm-2.0-500m-pytorch"
+  hparams = TimesFmHparams(
+    backend=device,
+    per_core_batch_size=32,
+    horizon_len=128,
+    num_layers=50,
+    use_positional_embedding=False,
+    context_len=192,
```
[the remark this answers is not in the document]
Comment (misha-chertushkin) [reply notebooks/finetuning_example.py:101]: Done 👍

File: notebooks/finetuning_example.py [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+  try:
+    if torch.cuda.is_available():
+      torch.cuda.set_device(rank)
+
+    os.environ["MASTER_ADDR"] = config.master_addr
+    os.environ["MASTER_PORT"] = config.master_port
+    if not torch.distributed.is_initialized():
+      torch.distributed.init_process_group(backend="nccl", world_size=world_size, rank=rank)
+
+    finetuner = TimesFMFinetuner(model, config, rank=rank)
```
[the remark this answers is not in the document]
Comment (misha-chertushkin) [reply notebooks/finetuning_example.py:231]: This is how pytorch handles the multi-GPU training in DDP regime: rank represents the ID of the process, which maps 1-to-1 to available gpu_ids. For example if you have 4 GPUS with ids=[0, 1, 2, 3], for each GPU there will be process created with relevant rank.

File: notebooks/finetuning_example.py [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+
+  results = return_dict.get("results", None)
+  print("\nFinetuning completed!")
+  return results
+
+
+if __name__ == "__main__":
+  try:
+    # single_gpu_example()  # Single GPU
+    multi_gpu_example()  # Multi-GPU
```
[the remark this answers is not in the document]
Comment (misha-chertushkin) [reply notebooks/finetuning_example.py:289]: Done :)

File: notebooks/finetuning_torch.py (hunk 2 above)
Comment (misha-chertushkin) [reply notebooks/finetuning_torch.py]: This is very good catch - I was reporting only the local loss. The gradient was properly done and the losses went down, but we only saw in wandb report the local loss of the respected process, and not the total sum of losses. I fixed it with using `all_reduce()` instead of `all_gather()`

File: notebooks/finetuning_torch.py (hunk 3 above)
Comment (misha-chertushkin) [reply notebooks/finetuning_torch.py]: Yes, this is very good catch. I have added 2 parameters: `log_every_n_steps` and `val_check_interval`. I also showed 2 examples how to use them :)

File: notebooks/finetuning_torch.py (hunk 1 above)
Comment (misha-chertushkin) [reply notebooks/finetuning_torch.py]: Done :) In order not to touch pytorch_patched_decoder.py I re-implemented quantile_loss on the side, and just add it to the total loss if quantile_loss=True in the config. Let me know if this approach looks good and it makes more sense to adjust pytorch_patched_decoder.py :)

File: src/finetuning/finetuning_torch.py [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+        """
+    x_context, x_padding, freq, x_future = [
+        t.to(self.device, non_blocking=True) for t in batch
+    ]
+
+    predictions = self.model(x_context, x_padding.float(), freq)
+    predictions_mean = predictions[..., 0]
+    last_patch_pred = predictions_mean[:, -1, :]
+
+    loss = self.loss_fn(last_patch_pred, x_future.squeeze(-1))
```
Comment (spMohanty) [inline src/finetuning/finetuning_torch.py:268]: I think before we calculate the loss, we should do something like:

```python
target_length = x_future.shape[1]
last_patch_pred = predictions_mean[:, -1, :target_length]
```

This handles the case when the target future (`x_future`) has a shorter time dimension than the model's output (i.e., say, when the horizon length in `x_future` is less than `output_patch_len`). Without this slice, we risk a shape mismatch between the predictions and the target. We should also apply similar slicing to the quantile predictions to ensure consistency.

This behavior appears to be handled correctly in the [[reference JAX implementation](https://github.com/google-research/timesfm/blob/9594c0618dec116e5006ef71a3d7f19630e00a0c/src/timesfm/patched_decoder.py#L536)](https://github.com/google-research/timesfm/blob/9594c0618dec116e5006ef71a3d7f19630e00a0c/src/timesfm/patched_decoder.py#L536).

**Comment by spMohanty** [comment]
The `horizon_len` parameter provided via `TimesFmHparams` is being altered internally by the `TimesFm` class. This behavior can cause unexpected errors in downstream tasks that rely on the user-specified horizon length.

For example, consider the following code:

```python
device = "cuda" if torch.cuda.is_available() else "cpu"
repo_id = "google/timesfm-2.0-500m-pytorch"
hparams = TimesFmHparams(
    backend=device,
    per_core_batch_size=32,
    horizon_len=256,
    num_layers=50,
    use_positional_embedding=False,
    context_len=192,  # Context length can be anything up to 2048 in multiples of 32
)

print(f"Intended horizon len: {hparams.horizon_len}")
tfm = TimesFm(hparams=hparams,
              checkpoint=TimesFmCheckpoint(huggingface_repo_id=repo_id))
print(f"Reported horizon len: {tfm._model_config.horizon_len}")
```

The output is:

```
Intended horizon len: 256
Reported horizon len: 128
```

**Review by rajatsen91 — APPROVED** [review]
